S3 (#1497)
hadley/adv-r: S3.Rmd

* Check the type of the base object and the types of each attribute.

> I'll illustrate these ideas by creating constructors for base classes[^base-constructors] that you're already familiar with. To start, lets make a constructor for the simplest S3 class: `Date`. A `Date` is just a double with "Date" as its `class` attribute, and no additional attributes. This makes for a very simple constructor:

Reviewer: How does this sound?

```suggestion
I'll illustrate these ideas by creating constructors for base classes[^base-constructors] that you're already familiar with. To start, lets make a constructor for the simplest S3 class: `Date`. A `Date` is just a double with a single attribute: its `class` is "Date". This makes for a very simple constructor:
```

new_factor(0:1, "a")
```

> Rather than encumbering the constructor, it's better to put complicated checks in a separate function. Doing so allows you to cheaply create new objects when you know that the values are correct, and easily re-use the checks in other places.

Reviewer: I prefer the previous wording. Can you please revert?

*   First, you should only ever write a method if you own the generic or the
    class. R will allow you to define a method even if you don't, but it is
    exceedingly bad manners. Instead, work with the author of either the 

>     generic or the class to add the method in his or her code.

Reviewer: I'm fine with singular they, so can you please revert?

Author: I also like it better. I think there are other instances of "his or her" in other chapters. I can do a global check after everything is merged so it's consistent throughout.
